sympy/core/expr.py

```diff
@@ -1980,15 +1985,39 @@ def as_coefficients_dict(self):
         0
         >>> (3*a*x).as_coefficients_dict()
         {a*x: 3}
+        >>> (3*a*x).as_coefficients_dict(x)
+        {x: 3*a}
 
         """
-        c, m = self.as_coeff_Mul()
-        if not c.is_Rational:
-            c = S.One
-            m = self
-        d = defaultdict(int)
-        d.update({m: c})
-        return d
+        if not syms:
+            d = defaultdict(list)
+            for ai in Add.make_args(self):
+                c, m = ai.as_coeff_Mul()
+                d[m].append(c)
+            for k, v in d.items():
+                if len(v) == 1:
+                    d[k] = v[0]
+                else:
+                    d[k] = Add(*v)
+            di = defaultdict(int)
+            di.update(d)
+            return di
```

Comment:
Why does this return a defaultdict? It's usually better not to because they can behave in unexpected ways.

Reply:
Because that is the advertised behavior of the default case (no symbols) per docstring:
```
        """Return a dictionary mapping terms to their Rational coefficient.
        Since the dictionary is a defaultdict, inquiries about terms which
        were not present will return a coefficient of 0. If an expression is
        not an Add it is considered to have a single term.
```
I'm not sure it is worth a deprecation.